_drafts/2020-09-17-fixing-an-n1-performance-problem-in-xaf-xpo-with-totally-undocumented-apis.md

```diff
@@ -0,0 +1,948 @@
+    public string Name { get => _Name; set => SetPropertyValue(nameof(Name), ref _Name, value); }
+
+    [Association, Aggregated]
+    public XPCollection<FasterOfferItemWithCQRS> OfferItems => GetCollection<FasterOfferItemWithCQRS>(nameof(OfferItems));
+}
+
+public class FasterOfferItemWithCQRS : BaseObject
+{
+    public FasterOfferItemWithCQRS(Session session) : base(session) { }
+
+    protected override void OnSaving()
+    {
+        //Mark the parent as saved, if we change
+        FasterOfferWithCQRS?.Save();
+        base.OnSaving();
+    }
+
+    protected override void OnDeleting()
+    {
+        //Mark the parent as saved, even if we are deleted
+        FasterOfferWithCQRS?.Save();
+        base.OnDeleting();
+    }
+
+    private int _Hours;
+    [Persistent("Hours")]
+    public int Hours { get => _Hours; set => SetPropertyValue(nameof(Hours), ref _Hours, value); }
+
+    private FasterOfferWithCQRS _FasterOfferWithCQRS;
+    [Persistent("FasterOfferWithCQRS"), Association]
+    public FasterOfferWithCQRS FasterOfferWithCQRS { get => _FasterOfferWithCQRS; set => SetPropertyValue(nameof(FasterOfferWithCQRS), ref _FasterOfferWithCQRS, value); }
+}
+
+[DefaultProperty(nameof(HourSum))]
+public class FasterOfferWithCQRSQuery : BaseObject
+{
+    public FasterOfferWithCQRSQuery(Session session) : base(session) { }
+
+    private FasterOfferWithCQRS _FasterOfferWithCQRS;
+    [Persistent("FasterOfferWithCQRS")]
+    [MemberDesignTimeVisibility(false)] //We don't want to see the data in the view, we only want the reference as a foreign key
+    //Save the reference for comparison. We also can use a unique index on this if needed, but we omit that for simplicity
+    public FasterOfferWithCQRS FasterOfferWithCQRS { get => _FasterOfferWithCQRS; set => SetPropertyValue(nameof(FasterOfferWithCQRS), ref _FasterOfferWithCQRS, value); }
+
+    private string _Name;
+    [Persistent("Name")]
+    public string Name { get => _Name; set => SetPropertyValue(nameof(Name), ref _Name, value); }
+
+    private int _HourSum;
+    [Persistent("HourSum")]
+    public int HourSum { get => _HourSum; set => SetPropertyValue(nameof(HourSum), ref _HourSum, value); }
+}
+```
+
+> **Important**: this only covers this very basic example, and highly depends on your configuration (e.g. SecuriySystem, Defered deletion etc.). You will need good testing on that to not get stale or out of sync.  
+Database side foreignkeys can help to mark stale objects, as well as triggers to handle that on a database level.  
+Also this *simple* technique only works if you are the only on that writes into the database.  
+I will cover mode advanced scenarios in the future, just let me know in the comments below.
+
+This will result in a slower write performance for the `Order` but a **huge** speed improvment on the `Query` table. I've set the `DataAccessMode` of the `Query` table to `DataView` and left the `DataAccessMode` for the `OrderTable` in `Client` mode:
```

Comment:
```suggestion
This will result in a slower write performance for the `Offer` but a **huge** speed improvement on the `Query` table. I've set the `DataAccessMode` of the `Query` table to `DataView` and left the `DataAccessMode` for the `OfferTable` in `Client` mode:
```